packages/openapi-v3/src/consolidate-schema.ts

```diff
@@ -0,0 +1,106 @@
+import compare from 'json-schema-compare';
+import _ from 'lodash';
+import {
+  ISpecificationExtension,
+  isSchemaObject,
+  OpenApiSpec,
+  PathsObject,
+  ReferenceObject,
+  SchemaObject,
+} from './types';
+
+let $schemaRefs: {
+  [schema: string]: SchemaObject | ReferenceObject;
+};
+let $paths: PathsObject;
+
+/*
+ *  Recursively search OpenApiSpec PathsObject for SchemaObjects with title property.
+ *  Move reusable schema bodies to #/components/schemas and replace with json pointer.
+ *  Handles collisions /w title, schema body pair comparision.
+ */
+export function ConsolidateSchemaObjects(spec: OpenApiSpec): OpenApiSpec {
+  $schemaRefs =
```

Comment:
We are avoiding global variables (shared state) in our code base. While they are safe to use as long as the function is synchronous, it makes it easy to accidentally introduce a race condition when a sync function changes to async and the developer forget about the shared state.

Our usual approach is to create some sort of a "context" object that can hold `$schemaRefs`, `$paths` and any other data. This context object is passed as a function argument to lower-level helpers.

Alternatively, you can also create a new class and move schemaRefs & paths to member (instance) properties, treating `this` as the context.

Reply:
I'll see if I can figure out your usual approach and implement.

Reply:
FYI - I created a story for adding the oaiEnhancerService in the rest module:
https://github.com/strongloop/loopback-next/issues/4380

Even before #4380 happen, I think you can still create the consolidation as a class that implements the enhancer. And inside `getApiSpec()`, temporarily apply it as:

```ts
// ...other code
const consolidationEnhancer = new ConsolidationEnhancer();
spec = consolidationEnhancer.modifySpec(spec);
// ...
```